Add Slots Documentation (#4527)
aws-amplify/docs: src/pages/console/uibuilder/slots.mdx

export const meta = {
  title: `Slots`,

>   description: `Add slots to Amplify Studio generated Figma-to-code components.  Used to support Nested Components or Collections in React code.`,

Reviewer: ```suggestion
  description: `Add slots to Amplify-generated Figma-to-code components.  Use this to support nested components or collections in React code.`,
```

};

Slots allow you to nest other components as React code within your Studio UI components.  You can use slots to create dynamically generated children, like Comments on a Post, or to replace a child element altogether.

## Adding a slot
First, you'll need a component.  
1. Log into Amplify and [navigate to Studio](/console/adminui/start/).

2. On the left-hand navigation bar, click UI Library

3. Select a Figma component from your UI Library that you've imported into Studio.  
*If you don't have any Figma components already, you can start with [Amplify's Figma UI file](/console/uibuilder/figmatocode/).*

> 4. Configure that component by clicking the Configure button in the upper right-hand corner.

Reviewer: ```suggestion
4. Configure that component by clicking the **Configure** button in the upper right-hand corner.
```

![Open your slot and click the gear icon to configure](/images/console/slots-console1.gif)

Next, you'll add a slot to this component.  

> 1. On the left-hand side, you'll see the Elements of your component.  Select a Figma Frame (![Figma frame icon](/images/console/figma-frame-icon.png)) within your component.  

Reviewer: ```suggestion
1. On the left-hand side, you'll see the elements of your component.  Select a Figma **Frame** (![Figma frame icon](/images/console/figma-frame-icon.png)) within your component.  
```

> 1. On the left-hand side, you'll see the Elements of your component.  Select a Figma Frame (![Figma frame icon](/images/console/figma-frame-icon.png)) within your component.  

Reviewer: the content in ![...] are accessibility labels for screenreaders. Please do a full pass on all of them, so it makes sense when the screen reader prompts the content.

  In this example, the "Area" frame is selected.

> 2. On the right-hand bar action bar, click the "Convert to a slot" button.  This will add a new component under the "Component properties" header.

Reviewer: ```suggestion
2. On the right-hand panel, click the "Convert to a slot" button. This will add a new prop to your UI component that. Any JSX element you pass into that prop will be rendered in the generated slot.
```

3. Optionally, change your your property name.  

  In this example, the property has been renamed "comments"

![Select your frame and add a slot](/images/console/slots-console2.gif)

<Callout warning>

Want to undo your slot creation?  

> Locate your slot in the Component properties, click the triple-dot icon, and click "Erase property" to remove the slot.

Reviewer: ```suggestion
Locate your slot in the **Component properties** (top-right corner), click the triple-dot icon, and click "Erase property" to remove the slot.
```

</Callout>

> ## Importing your component

Reviewer: ```suggestion
## Render the component with slots in React code
```

I don't like my suggestion either but something that points it clearer to the React component + slots would be good.

> Once you've added a slot to your component, click the "Get component code" button at the bottom of the screen to see instructions on the next steps.  

Reviewer: ```suggestion
Once you've added a slot to your component, click the **Get component code** button at the bottom of the screen to see instructions on the next steps.  
```

1. Copy the `amplify pull` command, and run it in your Terminal

> 2. Copy the import code and paste it in your App.js file

Reviewer: ```suggestion
2. Copy the import code and paste it in your React app code
```

> 3. Lastly, render the component by calling it in JSX

Reviewer: ```suggestion
3. Lastly, render the component
```

```jsx
/*Import your component*/
import { 
  Ampligram 
} from './ui-components'

function App() {
  return (
    /*Add your component below*/
    <div style={styles.post}>
      <Ampligram  /> 
    </div>
  );
}

const styles = {
  post: { width: 400, margin: '0 auto', padding: 20},
}

export default App;

```

> Here's how the code above would render in your app.  Some minor styling has been added to help with visibility.

Reviewer: ```suggestion
Here's how the code above would render in your app. *Some minor styling has been added to help with visibility.*
```

![Imported component with a slot](/images/console/slots-base.png)

> ## Using your slot

Reviewer: Potentially... "Using slots to replace child components"

> To use the slot, pass a child component as a property of the parent component, using the prop name you picked earlier.  Then, the content you pass will be rendered as a child of the component.  

Reviewer: ```suggestion
To use the slot, pass a child component as a property of the parent component, using the prop name you picked earlier. Then, the content you pass will be rendered as a child of the component.  
```

````jsx
import { 
  Ampligram 
} from '.ui-components'

function App() {
  return (
    <div style={styles.post}>

>       <h1>Simple Slot Children</h1>

Reviewer: I'm not sure we need the high-level div element. As long as the screenshots are "centered" I think we're good. The additional styles + elements might be distracting from the core of the guide.

      <Ampligram comments={

>         /* Pass your content into the slot */

Reviewer: Does this work? I didn't know that comments could be in-lined like this. Super cool!

Author: @renebrandel it worked in the local preview, so I'm expected it to work when published too!

        <div style={styles.comment}> 
          <h1>Hi mom!</h1>
          <p>Thanks for checking out my app</p>
        </div>
      }/>
    </div>
  );
}

const styles = {
  post: { width: 400, margin: '0 auto', padding: 20},
  comment: { paddingLeft: 30, textAlign: 'left'},
}

export default App;

> 

Reviewer: remove empty line

````

Here's how the code above would render in your app.  We've added a little more styling to the comments to help with display.

![Simple slot children](/images/console/slots-simple-children.png)

> ## Adding slots to collections

Reviewer: ```suggestion
## Adding slots to collections
```

In the same veins as above. Potentially an "H3": "Using slots to nest collections"

> Any component can be [converted to a collection](/console/uibuilder/collections/) and bound to data, and that includes components with slots.  With a collection, you can [extend your component via code](/console/uibuilder/override/) using the `overrideItems` prop to generate unique content within each component slot.  

Reviewer: Please do a full pass on all content and remove "double spacing" after period. Just like to keep it consistent across the docs with `. More content.` instead of `.  More content.`

Here, we have our Ampligram collection mapped to a data model called Post.  The Post model has a field called Comment, which contains an array of all the Comments associated with this Post.  We are then mapping the Comments for this Post to the "comments" slot we created.  

> Creating dynamic child components is where Slots get very useful - you can even [pass another collection](/console/uibuilder/override/#nesting-collections) into this slot. 

Reviewer: ```suggestion
Dynamically rendering child components is where Slots get very useful - you can even [pass another collection](/console/uibuilder/override/#nesting-collections) into this slot. 
```

The code below will render an Ampligram for each Post in the AmpligramCollection.  Then, the Comments for each Post will render in the "comments" slot.  

In this example, the first Post has 2 comments, but the second post has none.

```jsx
import { 
  AmpligramCollection 
} from './ui-components'

function App() {
  return (
    <div style={styles.post}>

>       <h1>Dynamic Slot Children</h1>

Reviewer: Same as above. This is potentially distracting from the rest of the guide.
